### PR TITLE
Add HDR support for D3D12 (rebased PR from MajorPainTheCactus)

### PR DESCRIPTION
**Pull request by twinaphex**
Rebased PR from https://github.com/libretro/RetroArch/pull/12916

@MajorPainTheCactus hopefully this will be alright? This is how I accomplished this BTW - do this on a branch -

```
git reset $(git merge-base master $(git rev-parse --abbrev-ref HEAD)) 
```

### CODE REVIEW

**Comment by twinaphex** [comment]
There are some build failures the buildbot is picking up on that I will try to resolve.

**Comment by twinaphex** [comment]
OK @MajorPainInTheCactus, the remaining build failures are with Mingw/Msys2 builds for Windows (x64 and x86). See here the logs - 

https://git.libretro.com/libretro/RetroArch/-/jobs/808778

https://git.libretro.com/libretro/RetroArch/-/jobs/808777

Error seems to be the same -

```
obj-unix/release/gfx/common/d3d12_common.o:d3d12_common.c:(.text+0xa3a): undefined reference to `IID_IDXGIOutput6'
```

**Comment by MajorPainTheCactus** [comment]
Ah I've never tried the UWP version. How do you compile and use it? Is there a page somewhere?

Other than that all looks good. I'm not sure how you got mingw/msys2 build issues I built both debug and release and all the checks passed when I created the pull request.

Is there somewhere I can look to see the build failures?

**Comment by MajorPainTheCactus** [comment]
I'd also love to use visual studio - I'm much more use to that than visual code which out of the box at least is quite poor (I'm sure there are a million extensions I can apply to make it better - lol)

**Comment by MajorPainTheCactus** [comment]
Apologies just saw theessage above showing the build jobs failing under Linux. I'll take a look and try to understand your changes when I get the chance - I'm working at the moment.

**Comment by twinaphex** [comment]
Hi @MajorPainTheCactus, the build/compilation issues on Mingw/msys2 Linux were actually fixed.

**Comment by twinaphex** [comment]
So basically @MajorPainTheCactus I think the only thing that I need some clarification on is those two comments, apart from that we should be ready to go. The Gitlab CI pipeline passes for all platforms now, and I have tested it on both MSVC (Visual Studio) and mingw/msys2, and it both works there.

File: gfx/common/dxgi_common.c [hunk 1]
```diff
@@ -485,4 +485,32 @@ bool dxgi_check_display_hdr_support(DXGIFactory factory, HWND hwnd)
...
+                  chain_handle, color_space,
+                  &color_space_support))
+            && ((color_space_support & 
+                  DXGI_SWAP_CHAIN_COLOR_SPACE_SUPPORT_FLAG_PRESENT) 
+               == DXGI_SWAP_CHAIN_COLOR_SPACE_SUPPORT_FLAG_PRESENT))
+      {
+         if (FAILED(DXGISetColorSpace1(chain_handle, color_space)))
+         {
+            RARCH_ERR("[DXGI]: Failed to set DXGI swapchain colour space\n");
+            /* TODO/FIXME/CLARIFICATION: Is this fall-through intentional?
```
[the remark this answers is not in the document]
Comment (twinaphex) [reply gfx/common/dxgi_common.c]: For now I've assumed this was an oversight and just added an early return here.

**Comment by twinaphex** [comment]
> Ah I've never tried the UWP version. How do you compile and use it? Is there a page somewhere?

You just open pkg/msvc-uwp and open either the VS2019 solution file or the VS2017 solution file (depending on whether you want to build with Visual Studio 2019 or Visual Studio 2017).

Make sure that 'Managed and Native' is set before building, otherwise some things might not work as expected.

image

> Is there somewhere I can look to see the build failures?

Yeah, on PRs that come from me or other internal contributors, a CI pipeline gets ran. You can see that with this notification:

'ci/gitlab/git.libretro.com Pending — Pipeline running on GitLab'

So right now the current CI pipeline being ran would be this - you can see build logs for all RetroArch platforms there.

**Comment by twinaphex** [comment]
I've found one issue so far with HDR enabled in D3D12 that does not happen with HDR disabled -

I cannot enable `shaders_slang/cel/MMJ_Cel_Shader.slangp` or `shaders_slang/cel/MMJ_Cel_Shader_MP.slangp` and see any graphics output. I can only see a non-black screen if I open the menu by toggling the menu while ingame, otherwise I just see a black screen.

Here is an example of what it looks like with the menu toggled -

image

With HDR disabled this works fine. Shaders in question can be found here -

https://github.com/libretro/slang-shaders/tree/master/cel

**Comment by MajorPainTheCactus** [comment]
@twinaphex I'll look at the above issues as soon as I get the chance. I've added support for HDR in the D3D11 driver tonight which has been added to the other pull request - yes I'm having a nightmare with git and GitHub - lol.

**Comment by twinaphex** [comment]
Cool. So far yeah the only issue I see is with that Cel shader, everything else works as before in HDR mode. If we can fix that, I think we could merge this PR. You can always publish a diff patch here in case you manage to fix it, that way I can apply it on top of this current PR and we can get it merged that way without you having to wrestle with Git/Github.

**Comment by MajorPainTheCactus** [comment]
Ah right yes I'#ve done that before with patches - seemed quite straight forward.  

Ok I'll look at the cel shader as soon as I get the chance.  Is this broken when HDR is off?  

Just so you are aware HDR will break aliasing around letters etc.  This is because we're changing the curve and over brightening/over darkening pixels beyond what the aliased colour should be.  The way to properly fix this is to render all the UI at paper white nits values i.e the top of the SDR range and then it should all look fine.  Currently we're rendering all the UI at center of the sun of nits (ok not quite) which is super bright and will be quite painful to look at for extended periods of time on very bright displays.  

I'll create a function that gives paper white values - it just needs to do a Reinhard tonemap (in the normal direction) to get the paper white value for the settings we're using.

Thats another point in the future the HDR setting page should really be done with calibration images as everybody's TV/monitor will be different and its difficult to know what is the best setting for their particular model.

**Comment by MajorPainTheCactus** [comment]
Apologies just saw your point that the cel shading isn't broken when HDR is off.  By the sounds of it hopefully this will be relatively simple to fix as its so obviously broken.

**Comment by MajorPainTheCactus** [comment]
> 
> 
> @MajorPainTheCactus So this is a question I have - should we still have an else block here where this gets performed if HDR is disabled?
> 
> ```
>       d3d12_resource_transition(
>             d3d12->queue.cmd,
>             d3d12->chain.renderTargets[d3d12->chain.frame_index],
>             D3D12_RESOURCE_STATE_RENDER_TARGET,
>             D3D12_RESOURCE_STATE_PRESENT);
> ```
> 
> Or was this done intentionally? Check line 1970 of gfx/drivers/d3d12.c on the right. (click on the comment icon in commit [dbcd1c8](https://github.com/libretro/RetroArch/commit/dbcd1c868a017a7c35d57e4fce9f74369620eede), it will take you right to the comment).

Yes you're right this should be outside the if block - good catch!

**Comment by twinaphex** [comment]
Alright, I'll make that change already then.

**Comment by MajorPainTheCactus** [comment]
I've done it already in my fixes in my other pull request along with the fix for the cel shader - not the sahder itself the blend mode being wrong on the HDR shader draw call

**Comment by MajorPainTheCactus** [comment]
We should be all good now?

**Comment by twinaphex** [comment]
OK, this one?

https://github.com/libretro/RetroArch/pull/12916

I can test that out beforehand and if all is well, then sure.

**Comment by twinaphex** [comment]
Hmmm, well, that PR still does not do the necessary changes that I made though, like moving the common DXGI code to dxgi_common.c and instead having all this code in d3d12_common.c and d3d11_common.c for instance.

**Comment by twinaphex** [comment]
I'm also missing the HAVE_DXGI_HDR ifdefs as well. It should still be possible to compile without baking HDR support in in some hypothetical scenario where this is necessary.

**Comment by twinaphex** [comment]
It also does not have the mingw msys2 unix buildfix -

https://github.com/libretro/RetroArch/pull/12917/commits/5a14336be0c7a4e19accc1f035518a5c3de5f3f9

Mind if we just use this PR as a base instead? I can backport what is remaining.

**Comment by MajorPainTheCactus** [comment]
Sorry all I mean was that my fixes went into that pull request not this pull request because I don't know how to get my fork to know about this pull request.  We should still use this one with all your changes in and I was hoping you could pull in the additional changes I made to fix the cel shader issue and support D3D11.
Is that not possible?

**Comment by twinaphex** [comment]
Ah @MajorPainTheCactus , OK then, I will do that. Thanks for the fixes/submissions so far.

BTW, it'd be awesome if you could come visit us on Discord if you'd like (my username on there is Autechre, you can just DM me in case you don't like speaking in public). Maybe we can try finding people who can help you with these Github trials you're having.

**Comment by twinaphex** [comment]
OK @MajorPainTheCactus, this seems good enough to merge at this point.

A couple of things I'd like you to look at for a future PR -

* Take a look at these TODO/FIXME notes - https://github.com/libretro/RetroArch/pull/12917/commits/52639aa3c4cab4e869eebdacb162ccdeba7356ef. They pop up when I run a Debug build with Visual Studio 2017/2019. They are Direct3D warnings so we might want to take care of them.

* Can you write up sublabels for these settings? We'd like to be able to explain these settings to our users better.
image

**Comment by MajorPainTheCactus** [comment]
Ok I'll definitely fix those things and write those tags. Sorry about all the problems it seems to have all stemmed from my initial issue of not realising I needed to rebase before pushing I'm also not sure what to do about only being able to have one pull request open at any one time. Anyway I'll close that pull request now.

**Pull request merged**